include/aws/common/array_list.h

```diff
@@ -0,0 +1,265 @@
+                                                                                                                       \
+    list->current_size = array_size;                                                                                   \
+    list->length = 0;                                                                                                  \
+    list->data = raw_array;                                                                                            \
+    list->alloc = 0;                                                                                                   \
+    return AWS_OP_SUCCESS;                                                                                             \
+}                                                                                                                      \
+                                                                                                                       \
+/**
+* Deallocates any memory that was allocated for this list, and resets list for reuse or deletion.
+*/                                                                                                                     \
+static inline void aws_##name##_array_list_clean_up(struct aws_##name##_array_list *list) {                            \
+                                                                                                                       \
+    if (list->alloc) {                                                                                                 \
+        aws_mem_release(list->alloc, (void *)list->data);                                                              \
+    }                                                                                                                  \
+                                                                                                                       \
+    list->current_size = 0;                                                                                            \
+    list->length = 0;                                                                                                  \
+    list->data = 0;                                                                                                    \
+    list->alloc = 0;                                                                                                   \
+}                                                                                                                      \
+                                                                                                                       \
+/**
+* Pushes a the dereferenced value of val onto the end of internal list->
+*/                                                                                                                     \
+static inline int aws_##name##_array_list_push_back(struct aws_##name##_array_list *list, T *val) {                    \
```

Comment:
`const T *`

Reply:
this one is complicated. That doesn't work when T is something like `void *` c const rules are weird.

Reply:
That means if we're passed in a `const void *` and pass it along to push_back it'll fail with a really weird error. Now I need to cast away the constness. There's some real hidden complexity here with this pseudo-templating.

We're not writing C++. We're writing C. When you have a generic datastructure in C, you use `void *`s and `memcpy`. You don't go and reimplement C++ templates in your header files, because strange non-idiomatic stuff like this happens.

Reply:
I'll redo it.